### ISSUE TITLE
False-positive: Invalid foreach argument

### ISSUE THREAD

**Issue by cautionbug** [kalessil/phpinspectionsea#1270]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Plugin         | Php Inspections (EA Extended) 3.0.12.2 |
| Language level | PHP 7.2 |

Current behaviour
--

Foreach argument inspection shows false-positive when argument is string-tested with other conditions:

```php
$foo = [];

// Works:
if (is_string($foo)) {
} else {
    // No inspection complaint
    foreach ($foo as $bar) { }
}

// False positive
if (is_string($foo) && true) {
} else {
    // Inspection: "Expected types: array or object, Actual: string"
    foreach ($foo as $bar) { }
}
```

Expected behaviour
--
When coupling `is_string()` with other required (`&&`) conditions (per example), inspection should know the argument cannot be a string.

This is probably challenging as it requires picking apart complex conditions (`A && B || C && is_string($foo)`) but maybe there's middle ground that if all logical operators are `&&` it's safe to assume the `else` is not a string? Perhaps an alternate inspection notice that the argument's type is not guaranteed based on conditions?

Environment details
---

PhpStorm 2019.1
Build #PS-191.6183.95, built on March 27, 2019
Licensed to Professional Computing Resources / Joe Theuerkauf
Subscription is active until April 28, 2019
JRE: 1.8.0_202-release-1483-b39 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Windows 10 10.0

**Comment by kalessil** [comment]
One of the IDE related reports is https://youtrack.jetbrains.com/issue/WI-37859, but it was similar where IDE type resolver doesn't handle complex conditions.

**Comment by cautionbug** [comment]
Thanks for the quick response!
